### PR TITLE
Simplify TransportShardBulkAction#performOnReplica

### PR DESCRIPTION
**Pull request by original-brownbear**
* Resolve TODO since 8.0 doesn't have to worry about pre 6.x nodes
* Remove test for removed method since the logic is now completely internal to `performOnReplica`

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java [hunk 1]
```diff
@@ -442,25 +393,20 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
...
-                    location = syncOperationResultOrThrow(operationResult, location);
-                    break;
-                default:
-                    throw new IllegalStateException("illegal replica item execution mode for: " + docWriteRequest);
+            final BulkItemResponse response = item.getPrimaryResponse();
+            final BulkItemResponse.Failure failure = response.getFailure();
+            final DocWriteResponse writeResponse = response.getResponse();
+            final long seqNum = failure == null ? writeResponse.getSeqNo() : failure.getSeqNo();
+            if (seqNum == SequenceNumbers.UNASSIGNED_SEQ_NO) {
+                continue;
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: can we assert failure is not null here?

**Review by dnhatn — APPROVED** [review]
Nice clean up! Thanks @original-brownbear.

I left a minor ask.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: @dnhatn appears we cannot do that (yet), that assertion trips BwC tests.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1
Jenkins run elasticsearch-ci/2
Jenkins run elasticsearch-ci/bwc

**Comment by original-brownbear** [comment]
@dnhatn just re-requested review to be safe here :) See above I couldn't add the assertion since it breaks BwC tests, that's not indicative of an issue is it?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java]: Just to add a little more detail here: the assertion trips on responses like:

```json
{"update":{"_index":"test_index","_type":"test_type","_id":"test_id_1","_version":2,"result":"noop","forced_refresh":true,"_shards":{"total":0,"successful":0,"failed":0},"get":{"_seq_no":3,"_primary_term":1,"found":true,"_source":{"bar":"foo"}},"status":200}}
```

in the BwC tests.

**Review by dnhatn — APPROVED** [review]
Thanks, @original-brownbear for an extra iteration. We can either remove or relax the assertion to account the cases where the result is Result.NOOP or Result.NOT_FOUND. I am good with either way.

**Comment by original-brownbear** [comment]
Thanks @dnhatn I went with the relaxed assertion now :)

**Pull request merged**